### ISSUE TITLE
Pointer-generator does not work on CPU

### ISSUE THREAD

**Issue by kylebgorman** [CUNY-CL/yoyodyne#58]
With `--accelerator cpu` (or not specifying since that's the default):

```
  File "/home/kbg/.miniconda3/lib/python3.9/site-packages/yoyodyne/models/pointer_generator.py", line 142, in decode_step
    attention_probs.scatter_add_(
RuntimeError: scatter(): Expected self.dtype to be equal to src.dtype
```

One of the argument tensors, I suppose, is on CPU and the other one is...IDK where. @Adamits, at your leisure: any insights here?

**Comment by Adamits** [comment]
In what environment does this happen? So far I cannot reproduce on my mac (2 GHz Quad-Core Intel Core i5). For pointer generator with or without features, nor by explicitly setting `--accelerator cpu`. I also tried either 1 or 2 layers.

**Comment by kylebgorman** [comment]
I am doing this:

0. Python 3.9.16
1. at HEAD
2. reinstalled from requirements.txt
3. Then my script is roughly:

yoyodyne-train \
    --experiment "${LANGUAGE}" \
    --train "${TRAIN}" \
    --dev "${DEV}" \
    --model_dir "${MODEL_DIR}" \
    --arch "${ARCH}" \
    --embedding_size 300 \
    --hidden_size 100 \
    --batch_size 32 \
    --dropout .3 \
    --gradient_clip_val 3 \
    --accelerator cpu \
    --max_epochs 60 \
    --log_every_n_step 20 \
    --gradient_clip_val 3 \
    --optimizer adam \
    --learning_rate .001 \
    --max_epochs 60 \
    --seed 49 \
    --precision 16

Actually, I think that's it, it's about precision. Commenting out that line
makes it go away. Can you replicate now?

**Comment by Adamits** [comment]
I think I need to read upon how they implement this. I can see that  (with `--precision 16`)`self.dtype` is `f32` on the modules. The LSTM encoders return `f32`, but the linear layer before deocding https://github.com/CUNY-CL/yoyodyne/blob/master/yoyodyne/models/pointer_generator.py#L503 returns `f16`.

**Comment by Adamits** [comment]
Another note, if before line 503 linked above, I do `source_attention_weights = source_attention_weights.type(self.dtype)`,  everything seems to work fine. So it seems this `source_attention_weights` f16 is the only issue, possibly only for scatter_add_?

**Comment by kylebgorman** [comment]
scatter_add_ is insisting on the matching dtypes, I think, so one just
needs to make its arguments dtype-conformant, right? I don't know
immediately whether you want to cast up or down the precision spectrum,
though.

**Comment by Adamits** [comment]
I think I can either 1) do this with with masking instead of `scatter_add_`, or 2) use `zeros_like` in pytorch so that `attention_weights` and `source_attention_weights` are the same dtype by default.

**Comment by Adamits** [comment]
The above two suggestions are not as straight forward as I was thinking. I have opted to simply generate the zeros tensor for  `attention_probs` with the dtype of `source_attention_weights`, which also solves the problem.

This makes me think we should probably check if we can reproduce the same results on CPU and GPU with 16bit precision.

**Comment by kylebgorman** [comment]
Closed in #59.
